Entities/ACDropShip.cpp

```diff
@@ -786,140 +649,69 @@ void ACDropShip::Update()
-		engineRot.SetDegAngle(m_Rotation.GetDegAngle() + m_MaxEngineAngle);
-	// Lateral control application
-	else
-		engineRot.SetDegAngle(m_MaxEngineAngle * m_LateralControl);
+    /////////////////////////////////
+    // Manage Attachable:s
+    Matrix engineRot = 0;
+    if (m_Rotation.GetDegAngle() > m_MaxEngineAngle) {
+        engineRot.SetDegAngle(m_Rotation.GetDegAngle() - m_MaxEngineAngle);
+    } else if (m_Rotation.GetDegAngle() < -m_MaxEngineAngle) {
+        engineRot.SetDegAngle(m_Rotation.GetDegAngle() + m_MaxEngineAngle);
+    } else {
+        // Lateral control application
+        engineRot.SetDegAngle(m_MaxEngineAngle * m_LateralControl);
+    }
 
-    if (m_pRThruster && m_pRThruster->IsAttached())
-    {
+    if (m_pRThruster && m_pRThruster->IsAttached()) {
         m_pRThruster->SetRotAngle(engineRot.GetRadAngle());
-        m_pRThruster->SetJointPos(m_Pos + m_pRThruster->GetParentOffset().GetXFlipped(m_HFlipped) * m_Rotation);
-        m_pRThruster->Update();
-
-        // Update the Atoms' offsets in the parent group
-        Matrix atomRot(FacingAngle(m_pRThruster->GetRotMatrix().GetRadAngle()) - FacingAngle(m_Rotation.GetRadAngle()));
-        m_pAtomGroup->UpdateSubAtoms(m_pRThruster->GetAtomSubgroupID(), m_pRThruster->GetParentOffset() - (m_pRThruster->GetJointOffset() * atomRot), atomRot);
-
-        m_Health -= m_pRThruster->CollectDamage();
     }
 
-    if (m_pLThruster && m_pLThruster->IsAttached())
-    {
+    if (m_pLThruster && m_pLThruster->IsAttached()) {
         m_pLThruster->SetRotAngle(engineRot.GetRadAngle());
-        m_pLThruster->SetJointPos(m_Pos + m_pLThruster->GetParentOffset().GetXFlipped(m_HFlipped) * m_Rotation);
-        m_pLThruster->Update();
-
-        // Update the Atoms' offsets in the parent group
-        Matrix atomRot(FacingAngle(m_pLThruster->GetRotMatrix().GetRadAngle()) - FacingAngle(m_Rotation.GetRadAngle()));
-        m_pAtomGroup->UpdateSubAtoms(m_pLThruster->GetAtomSubgroupID(), m_pLThruster->GetParentOffset() - (m_pLThruster->GetJointOffset() * atomRot), atomRot);
-
-        m_Health -= m_pLThruster->CollectDamage();
     }
 
     // Auto balancing with the up thrusters
-
-    if (m_pURThruster && m_pURThruster->IsAttached() && m_pULThruster && m_pULThruster->IsAttached())
-    {
-        if (m_AutoStabilize)
-        {
+    if (m_pURThruster && m_pURThruster->IsAttached() && m_pULThruster && m_pULThruster->IsAttached()) {
+        if (m_AutoStabilize) {
             // Use a PD-controller for balance
-            float change = 0.9 * m_AngularVel + 0.8 * m_Rotation.GetRadAngle();
-            if (change > 0.2)
-            {
-                if (!m_pURThruster->IsEmitting())
+            float change = 0.9F * m_AngularVel + 0.8F * m_Rotation.GetRadAngle();
+            if (change > 0.2F) {
+                if (!m_pURThruster->IsEmitting()) {
                     m_pURThruster->TriggerBurst();
+                }
                 m_pURThruster->EnableEmission(true);
```

Comment:
This was just reformatting of Data's code. But yeah, I guess the the point is that if there's too much of an angle, it enables the relevant thruster to straighten out right away. Remember that, right below, if there's not too much of an angle emissions are disabled. Essentially this pid controller stuff completely takes control of the emitter and mostly ignores its properties so it can work in short bursts for effective stabilization.

Reply:
Seems weird to me but if you say it's ok, then ok.